source/kotlin/src/main/kotlin/io/github/cosinekitty/astronomy/Main.kt

```diff
@@ -201,84 +290,124 @@ class Astronomy {
-            return -2.79 + 1.494119 * u - 0.0598939 * u2 + 0.0061966 * u3 - 0.000197 * u4;
+            u = y - 1900.0
+            u2 = u * u; u3 = u * u2; u4 = u2 * u2
+            return -2.79 + 1.494119 * u - 0.0598939 * u2 + 0.0061966 * u3 - 0.000197 * u4
         }
         if (y < 1941.0) {
-            u = y - 1920.0;
-            u2 = u * u; u3 = u * u2;
-            return 21.20 + 0.84493 * u - 0.076100 * u2 + 0.0020936 * u3;
+            u = y - 1920.0
+            u2 = u * u; u3 = u * u2
+            return 21.20 + 0.84493 * u - 0.076100 * u2 + 0.0020936 * u3
         }
         if (y < 1961) {
-            u = y - 1950;
-            u2 = u * u; u3 = u * u2;
-            return 29.07 + 0.407 * u - u2 / 233 + u3 / 2547.0;
+            u = y - 1950
+            u2 = u * u; u3 = u * u2
+            return 29.07 + 0.407 * u - u2 / 233 + u3 / 2547.0
         }
         if (y < 1986.0) {
-            u = y - 1975.0;
-            u2 = u * u; u3 = u * u2;
-            return 45.45 + 1.067 * u - u2 / 260 - u3 / 718.0;
+            u = y - 1975.0
+            u2 = u * u; u3 = u * u2
+            return 45.45 + 1.067 * u - u2 / 260 - u3 / 718.0
         }
         if (y < 2005) {
-            u = y - 2000;
-            u2 = u * u; u3 = u * u2; u4 = u2 * u2; u5 = u2 * u3;
-            return 63.86 + 0.3345 * u - 0.060374 * u2 + 0.0017275 * u3 + 0.000651814 * u4 + 0.00002373599 * u5;
+            u = y - 2000
+            u2 = u * u; u3 = u * u2; u4 = u2 * u2; u5 = u2 * u3
+            return 63.86 + 0.3345 * u - 0.060374 * u2 + 0.0017275 * u3 + 0.000651814 * u4 + 0.00002373599 * u5
         }
         if (y < 2050.0) {
-            u = y - 2000.0;
-            return 62.92 + 0.32217 * u + 0.005589 * u * u;
+            u = y - 2000.0
+            return 62.92 + 0.32217 * u + 0.005589 * u * u
         }
         if (y < 2150.0) {
-            u = (y - 1820.0) / 100.0;
-            return -20.0 + 32.0 * u * u - 0.5628 * (2150.0 - y);
+            u = (y - 1820.0) / 100.0
+            return -20.0 + 32.0 * u * u - 0.5628 * (2150.0 - y)
         }
 
         /* all years after 2150 */
-        u = (y - 1820.0) / 100.0;
-        return -20.0 + (32.0 * u * u);
+        u = (y - 1820.0) / 100.0
+        return -20.0 + (32.0 * u * u)
+    }
+
+    internal fun universalTime(tt: Double): Double {
+        // This is the inverse function of TerrestrialTime.
+        // This is an iterative numerical solver, but because
+        // the relationship between UT and TT is almost perfectly linear,
+        // it converges extremely fast (never more than 3 iterations).
+
+        // dt = tt - ut
+        var dt = terrestrialTime(tt) - tt
+        while (true) {
+            val ut = tt - dt
+            val ttCheck = terrestrialTime(ut)
+            val err = ttCheck - tt
+            if (err.absoluteValue < 1.0e-12) return ut
+            dt += err
+        }
+    }
+}
+
+internal data class TerseVector(val x: Double, val y: Double, val z: Double) {
```

Comment:
C# port uses stack allocated `struct` but this uses heap allocated `class`, there is [value class](https://kotlinlang.org/docs/inline-classes.html#members) which isn't possible here but in future the language will be address that hopefully.

Addition of `data` here means the language will add functionalities such as `equals` to [data classes](https://kotlinlang.org/docs/data-classes.html).

Reply:
That is helpful to know. I was wondering about support for value-type `struct` like C# has, mainly to reduce pressure on the garbage collector. However, I will not worry about optimizing things until (a) everything is working and tested, (b) there is a way to measure and compare execution times objectively, and (c) there is a demonstrated need for better efficiency. In the words of Donald Knuth: "Premature optimization is the root of all evil."

Reply:
Excellent 😊